Store filterable tags set as class variable. Add method to (re)load the filterable tags set.

rdswift opened the pull request:
Here is the sort of thing that I was referring to in https://github.com/metabrainz/picard/pull/2668#discussion_r2134314980.

This way, there is a method that can be called by the (future) code that registers (or deregisters) tags from plugins.  All that needs to happen is to call `Filter.load_filterable_tags(force=True)` from the (future) code, and that will force a reload of the filterable tags.

I suspect that there may be some additional optimization available by creating and storing the filter selection dialog widget in a class level variable as well, so that it is only built when when the `Filter.load_filterable_tags()` method runs and actually updates the `Filter.filterable_tags` set.  I haven't looked into this, but can if you want me to help with it.

I hope you don't mind these kind of code suggestions.  You've done a great job so far with this whole filtering project, and I'm just trying to help out in some areas of the code that I may be more familiar with.  Also, I've been working with @zas and @phw on Picard for a few years so I'm starting to understand what they're expecting when they request changes.

jpmsousa03 committed:
PICARD-150: Support for internal search/filter of the main two panes

This commit introduces a "Find" bar, accessible via Ctrl+F,
allowing users to search and filter songs directly within
the editing views. The Find bar supports filtering by tags,
filenames, and file paths, helping users quickly narrow down
results. Includes customizable filter options to improve
search precision and user workflow.
Co-authored-by: Francisco Lisboa <franciscolisboa@tecnico.ulisboa.pt>

francescouteiro committed:
fix: rebase & headers

During the rebase process, some imports had to be changed, due to the recent modification to the project's structure.

jpmsousa03 committed:
feat: find option in view menu

jpmsousa03 committed:
fix: find box gains focus based on the currently focused view

jpmsousa03 committed:
feat: add internationalization support for find UI elements

jpmsousa03 committed:
fix: update filter button text to show only single selected filter for better readability

jpmsousa03 committed:
fix: update filters to display display_names and translations

jpmsousa03 committed:
fix: removed unused import

rdswift committed:
Refactor to allow testing button text generation.  Fix tests.

rdswift committed:
Fix imports to pass isort test.

jpmsousa03 committed:
fix: Removed hidden tags from filter, and adressed issues with file_filters

jpmsousa03 committed:
fix: removed unnecessary _( for display name

jpmsousa03 committed:
fix: convert text to lower inside _filter_tree_items function

rdswift committed:
Refactor to accommodate testing.  Update tests.

jpmsousa03 committed:
fix: mock import

jpmsousa03 committed:
fix: changed _name to str

jpmsousa03 committed:
refactor: rename find toolbar action and status tip

jpmsousa03 committed:
fix: QSplitter::replaceWidget: Index -1 out of range

jpmsousa03 committed:
refactor: rename valid_tags to filterable_tags and update related methods

jpmsousa03 committed:
fix: filterable_tag_names return value to a set

jpmsousa03 committed:
refactor: update filter button label logic in FindBox

jpmsousa03 committed:
refactor: remove find icon from FindBox

jpmsousa03 committed:
refactor: rename Find to Filter

jpmsousa03 committed:
refactor: rename test method to reflect new class name

rdswift committed:
Use OrderedDict to avoid manual sorting when creating file filters

rdswift committed:
Rename `test_ui_find.py`.  Fix filterable tags generator test.

rdswift committed:
Make filter bar display a persistent setting to restore on startup

rdswift committed:
Fix tests

jpmsousa03 committed:
update comment to reflect new function name

rdswift committed:
Fix filterable tags generator to use `is_filterable` only

rdswift committed:
Add `filter_key` attribute to `TagVar` class

rdswift committed:
Revert "Add `filter_key` attribute to `TagVar` class"

This reverts commit 85346e4a8f197dca106ef0a38579b667e4888f61.

phw committed:
Merge pull request #2671 from rdswift/tag_filter_corrections

Tag filter corrections

rdswift committed:
Use `filterable_tag_names()` to provide all filter items:

- Cherry pick commit to fix `filterable_tag_names()` generator.
- Remove special `filename` and `filepath` filter items.
- Remove filter selection header sections.
- Remove unnessary `str()` call.
- Update `make_button_text()` method.
- Perform file filtering based on `~filename` and `~filepath` filters.
- Update tests.

rdswift committed:
Add tooltips to filter items.

rdswift committed:
Store filterable tags set as class variable. Add method to regenerate.

jpmsousa03 commented:
> I hope you don't mind these kind of code suggestions. You've done a great job so far with this whole filtering project, and I'm just trying to help out in some areas of the code that I may be more familiar with. Also, I've been working with @zas and @phw on Picard for a few years so I'm starting to understand what they're expecting when they request changes.

Not at all! I honestly appreciate it very much.

jpmsousa03 commented:
> I suspect that there may be some additional optimization available by creating and storing the filter selection dialog widget in a class level variable as well, so that it is only built when when the `Filter.load_filterable_tags()` method runs and actually updates the `Filter.filterable_tags` set. I haven't looked into this, but can if you want me to help with it.

I believe this is possible. However It would be necessary to also be updatable by a pluggin change "request" (the checkboxes), so I think this should be done before this merge (as I think one call should both update the filterlist and reload checkboxes).

rdswift commented:
> I think this should be done before this merge (as I think one call should both update the filterlist and reload checkboxes).

I agree.  Note that this PR is really just a starting point so far.  For example, a change in the `Filter.filterable_tags` set should also remove any items from each instance's `self.selected_filters` list that don't appear in the updated `Filter.filterable_tags` set (and re-apply the filter if the `self.selected_filters` list changes).  Achieving that could be a bit tricky.

rdswift committed:
Reapply filters if filterable tags reloaded.  Add testing option.

rdswift commented:
This commit adds some initial code for updating the filters when the `Filter.load_filterable_tags()` method is called.

1. Updates the `Filter.filterable_tags` set.
2. Removes any items in the `self.selected_filters` list that are no longer in the `Filter.filterable_tags` set.  Applied to all instances.
3. Reapplies the filters if the `self.selected_filters` list has changed.  Applied to all instances.

I've also included an option to include a button to test the functionality.  Clicking the test button toggles between including an additional test filterable tag and resetting to the tags provided by the `filterable_tag_names()` function.  This should be disabled (or more appropriately remove the test code entirely) once all development and testing is complete.

rdswift committed:
Avoid multiple builds of the filter selection dialog:

- Store the dialog as an instance variable.
- Build the dialog during initialization.
- Rebuild the dialog after reloading the filterable tags, and only if
the filterable tags list has changed.
- Remove unnecessary empty `self.selected_filters` check.

rdswift commented:
This commit attempts to avoid multiple builds of the filter selection dialog:

- Store the dialog as an instance variable.
- Build the dialog during initialization.
- Rebuild the dialog after reloading the filterable tags, and only if the filterable tags list has changed.

rdswift committed:
Add trigger to re-apply the filters after a file move

rdswift committed:
Add `suspended` flag to `Filter`.  Add context manager for filtering.

rdswift committed:
Remove testing button and code.

rdswift committed:
Use `set()` for selected filters

rdswift commented:
@jpmsousa03,

I've removed the test button and code, and have merged in the code to re-apply the filters after a drag & drop or adding a file to the cluster pane.  This also includes the context manager code to only re-appy the filter once when all added items have been processed.  I think this PR is ready to merge.  Please have a look.

zas committed:
Generic suspend while loading mechanism

- it can be extended using `register_suspend_while_loading()`
- it removes the need to call `MainWindow.set_sorting()` directly (it was a leftover when we implemented the context manager)

phw reviewed (approved):
This looks fine to me. With this we are all prepared for having this tied into the plugin system. 

No idea why CI was failing with the git fetch, but seems unrelated.

rdswift commented:
> No idea why CI was failing with the git fetch, but seems unrelated.

I wonder if @jpmsousa03's fork doesn't have the tags?  In any event, it does seem unrelated to the code changes.

rdswift committed:
Use `suspend_while_loading` to manage re-applying filters.

rdswift committed:
Merge branch 'search_feature' into search_feature

rdswift commented:
Rebased and modified to use the `suspend_while_loading` functionality from https://github.com/metabrainz/picard/pull/2672.

rdswift committed:
Use `set()` for all `selected_filters` references

rdswift committed:
Revert changes on `cancel`.  Add button to clear all selected filters.

rdswift committed:
Remove unneeded items from tests

rdswift commented:
Added code to revert any filter selection changes when `Cancel` is selected (bug) and added a button to clear all selected filters.

rdswift committed:
Merge remote-tracking branch 'jpmsousa03/search_feature' into search_feature

rdswift committed:
Refactor as per suggestion by zas and add more filter tests

jpmsousa03 commented:
All the code changes look good and are working well, great job, @rdswift! Since @phw also approved, I think this can be merged. I'll go ahead and update the "Clear All" button in the main PR.
